substrate/client/consensus/beefy/src/lib.rs

```diff
@@ -594,30 +630,54 @@ pub async fn start_beefy_gadget<B, BE, C, N, P, R, S, AuthorityId>(
 			is_authority,
 		);
 
-		match futures::future::select(
-			Box::pin(worker.run(&mut block_import_justif, &mut finality_notifications)),
-			Box::pin(on_demand_justifications_handler.run()),
-		)
-		.await
-		{
-			// On `ConsensusReset` error, just reinit and restart voter.
-			futures::future::Either::Left(((error::Error::ConsensusReset, reuse_comms), _)) => {
-				error!(target: LOG_TARGET, "🥩 Error: {:?}. Restarting voter.", error::Error::ConsensusReset);
-				beefy_comms = reuse_comms;
-				continue;
-			},
-			// On other errors, bring down / finish the task.
-			futures::future::Either::Left(((worker_err, _), _)) => {
-				error!(target: LOG_TARGET, "🥩 Error: {:?}. Terminating.", worker_err)
+		futures::select! {
+			result = worker.run(&mut block_import_justif, &mut finality_notifications).fuse() => {
+				match result {
+					(error::Error::ConsensusReset, reuse_comms) => {
+						error!(target: LOG_TARGET, "🥩 Error: {:?}. Restarting voter.", error::Error::ConsensusReset);
+						beefy_comms = reuse_comms;
+						continue;
+					},
+					(err, _) => {
+						error!(target: LOG_TARGET, "🥩 Error: {:?}. Terminating.", err)
+					}
+				}
 			},
-			futures::future::Either::Right((odj_handler_err, _)) => {
-				error!(target: LOG_TARGET, "🥩 Error: {:?}. Terminating.", odj_handler_err)
+			odj_handler_error = on_demand_justifications_handler.run().fuse() => {
+				error!(target: LOG_TARGET, "🥩 Error: {:?}. Terminating.", odj_handler_error)
 			},
-		};
+			_ = &mut transformer => {
+				error!(target: LOG_TARGET, "🥩 Finality notification transformer task has unexpectedly terminated.");
+			}
```

Comment:
That would be an option, true. However I think the impact of this additional future is pretty slim, since we are just using intra-task concurrency here. If we were to go for this route, we would need to re-initialize the transformer in case of a consensus reset where we start the loop again from the beginning. Also we would need to do some kind of "hand-over" for the streams to make absolutely sure that we don't miss any finality notifications. Since this raises the complexity I opted for the most simple solution here.